Introduce reset implementations for accumulators

timo0 opened the pull request:
Fix #6536

Implements for:
* DefaultDoubleAccumulator
* DefaultLongAccumulator
* LongAccumulator

Some house keeping, and tests

netopyr commented:
My idea was to make a couple more improvements when working on this PR:

- One should be able to provide a `LongFunction` respectively an `IntFunction` as an initializer. It gets the current value and returns the value the accumulator should be reset to. The possibility of providing an `IntSupplier` to `IntegerAccumulator` should be deprecated and removed in a future release.
- Unfortunately, `DefaultLongAccumulator` needs to use an `AtomicLong` internally. (Calling reset() on `java.util.concurrent.atomic.LongAccumulator` is problematic here.) An alternative could be to create a new instance of  `java.util.concurrent.atomic.LongAccumulator` on every reset.
- Switching from `throwArgNull()` to `Objects.requireNonNull()` should be done for all classes in the metrics package at once, IMO. Otherwise, you end up with some classes throwing IAE and others NPE until this gets cleaned up. Consistency is crucial. Also, the javadocs need to be updated.
- If you prefer to switch to NPEs partially, you must ensure that at least all implementations of `Metric.get(ValueType)` behave the same.

timo0 commented:
> * One should be able to provide a `LongFunction` respectively an `IntFunction` as an initializer. It gets the current value and returns the value the accumulator should be reset to. The possibility of providing an `IntSupplier` to `IntegerAccumulator` should be deprecated and removed in a future release.
> * Unfortunately, `DefaultLongAccumulator` needs to use an `AtomicLong` internally. (Calling reset() on `java.util.concurrent.atomic.LongAccumulator` is problematic here.) An alternative could be to create a new instance of  `java.util.concurrent.atomic.LongAccumulator` on every reset.

Will be done in separate issue


> * If you prefer to switch to NPEs partially, you must ensure that at least all implementations of `Metric.get(ValueType)` behave the same.

Agree and reverted...

> * Switching from `throwArgNull()` to `Objects.requireNonNull()` should be done for all classes in the metrics package at once, IMO. Otherwise, you end up with some classes throwing IAE and others NPE until this gets cleaned up. Consistency is crucial. Also, the javadocs need to be updated.

Will do in a separate PR too

poulok commented:
> * Unfortunately, `DefaultLongAccumulator` needs to use an `AtomicLong` internally. (Calling reset() on `java.util.concurrent.atomic.LongAccumulator` is problematic here.) An alternative could be to create a new instance of  `java.util.concurrent.atomic.LongAccumulator` on every reset.

> Will be done in separate issue

I don't think doing this later is an option. The javadoc for `java.util.concurrent.atomic.LongAccumulator#reset()` says that it is only effective is there are no other threads making concurrent updates. We do not guarantee this in our code.

```public void reset()
Resets variables maintaining updates to the identity value. This method may be a useful alternative to creating a new updater, but is only effective if there are no concurrent updates. Because this method is intrinsically racy, it should only be used when it is known that no threads are concurrently updating.
```

artemananiev reviewed (approved):
Looks fine to me

hendrikebbers commented on platform-sdk/swirlds-common/src/main/java/com/swirlds/common/metrics/DoubleAccumulator.java:
```diff
@@ -132,10 +135,12 @@ private Config(
                 final String unit,
                 final String format,
                 final DoubleBinaryOperator accumulator,
+                final DoubleSupplier initializer,
```
nullable annotation

hendrikebbers commented on platform-sdk/swirlds-common/src/main/java/com/swirlds/common/metrics/DoubleAccumulator.java:
```diff
@@ -180,6 +206,15 @@ public DoubleBinaryOperator getAccumulator() {
+         *
+         * @return the initializer
+         */
+        public DoubleSupplier getInitializer() {
```
null annotation for result

hendrikebbers commented on platform-sdk/swirlds-common/src/main/java/com/swirlds/common/metrics/DoubleAccumulator.java:
```diff
@@ -213,7 +255,35 @@ public double getInitialValue() {
+         * 		the initializer
+         * @return a new configuration-object with updated {@code initializer}
+         */
+        public DoubleAccumulator.Config withInitializer(final DoubleSupplier initializer) {
```
null annotations for param and result

hendrikebbers commented on platform-sdk/swirlds-common/src/main/java/com/swirlds/common/metrics/LongAccumulator.java:
```diff
@@ -131,10 +135,12 @@ private Config(
                 final String unit,
                 final String format,
                 final LongBinaryOperator accumulator,
+                final LongSupplier initializer,
```
nullable annotation

hendrikebbers commented on platform-sdk/swirlds-common/src/main/java/com/swirlds/common/metrics/LongAccumulator.java:
```diff
@@ -179,6 +206,15 @@ public LongBinaryOperator getAccumulator() {
+         *
+         * @return the initializer
+         */
+        public LongSupplier getInitializer() {
```
null annotation

hendrikebbers commented on platform-sdk/swirlds-common/src/main/java/com/swirlds/common/metrics/LongAccumulator.java:
```diff
@@ -191,7 +227,35 @@ public LongBinaryOperator getAccumulator() {
+         * 		the initializer
+         * @return a new configuration-object with updated {@code initializer}
+         */
+        public LongAccumulator.Config withInitializer(final LongSupplier initializer) {
```
null annotations

hendrikebbers commented on platform-sdk/swirlds-common/src/main/java/com/swirlds/common/metrics/atomic/AtomicDouble.java:
```diff
@@ -100,4 +103,45 @@ public final boolean compareAndSet(double expectedValue, double newValue) {
+     * @param accumulatorFunction the accumulator function
+     * @return the updated value
+     */
+    public final double accumulateAndGet(double x, DoubleBinaryOperator accumulatorFunction) {
```
null annotation

hendrikebbers commented on platform-sdk/swirlds-common/src/main/java/com/swirlds/common/metrics/atomic/AtomicDouble.java:
```diff
@@ -100,4 +103,45 @@ public final boolean compareAndSet(double expectedValue, double newValue) {
+     * @param updateFunction the update function
+     * @return the updated value
+     */
+    public final double updateAndGet(DoubleUnaryOperator updateFunction) {
```
null annotation

hendrikebbers commented on platform-sdk/swirlds-common/src/main/java/com/swirlds/common/metrics/platform/DefaultDoubleAccumulator.java:
```diff
@@ -20,39 +20,44 @@
-        this.initialValue = config.getInitialValue();
+        this.accumulator = config.getAccumulator();
+        final double initialValue = config.getInitialValue();
+        this.initializer = config.getInitializer() != null ? config.getInitializer() : () -> initialValue;
```
Since config is not a record 2 calls to `config.getInitializer()` could result in 2 different values. in worst case the second call could return `null`. Therefore I would refactor this to only have 1 call to `config.getInitializer()`

hendrikebbers commented on platform-sdk/swirlds-common/src/main/java/com/swirlds/common/metrics/platform/DefaultLongAccumulator.java:
```diff
@@ -22,37 +22,42 @@
-        this.initialValue = config.getInitialValue();
+        this.accumulator = config.getAccumulator();
+        final long initialValue = config.getInitialValue();
+        this.initializer = config.getInitializer() != null ? config.getInitializer() : () -> initialValue;
```
Since config is not a record 2 calls to config.getInitializer() could result in 2 different values. in worst case the second call could return null. Therefore I would refactor this to only have 1 call to config.getInitializer()

hendrikebbers commented on platform-sdk/swirlds-common/src/main/java/com/swirlds/common/metrics/atomic/AtomicDouble.java:
```diff
@@ -100,4 +104,46 @@ public final boolean compareAndSet(double expectedValue, double newValue) {
+     * @param delta the value to add
+     * @return the updated value
+     */
+    public final double addAndGet(double delta) {
```
```suggestion
    public final double addAndGet(final double delta) {
```

hendrikebbers commented on platform-sdk/swirlds-common/src/main/java/com/swirlds/common/metrics/atomic/AtomicDouble.java:
```diff
@@ -100,4 +104,46 @@ public final boolean compareAndSet(double expectedValue, double newValue) {
+     * @param accumulatorFunction the accumulator function
+     * @return the updated value
+     */
+    public final double accumulateAndGet(double x, @NonNull DoubleBinaryOperator accumulatorFunction) {
```
```suggestion
    public final double accumulateAndGet(final double x, @NonNull final DoubleBinaryOperator accumulatorFunction) {
```

hendrikebbers replied:
+ I do not like the name `x` for a param

hendrikebbers commented on platform-sdk/swirlds-common/src/main/java/com/swirlds/common/metrics/atomic/AtomicDouble.java:
```diff
@@ -100,4 +104,46 @@ public final boolean compareAndSet(double expectedValue, double newValue) {
+     * @param updateFunction the update function
+     * @return the updated value
+     */
+    public final double updateAndGet(@NonNull DoubleUnaryOperator updateFunction) {
```
```suggestion
    public final double updateAndGet(@NonNull final DoubleUnaryOperator updateFunction) {
```

hendrikebbers commented on platform-sdk/swirlds-common/src/main/java/com/swirlds/common/metrics/atomic/AtomicDouble.java:
```diff
@@ -100,4 +104,46 @@ public final boolean compareAndSet(double expectedValue, double newValue) {
+     * @param accumulatorFunction the accumulator function
+     * @return the updated value
+     */
+    public final double accumulateAndGet(final double updateValue, @NonNull DoubleBinaryOperator accumulatorFunction) {
```
```suggestion
    public final double accumulateAndGet(final double updateValue, @NonNull final DoubleBinaryOperator accumulatorFunction) {
```
